Save multiple datasets into one hdf5 file.

JWichmann opened the pull request:
This pull request adds the possibility to save multiple datasets in one hdf5 file.
The change is implemented such that old code is not affected.
Note: Autoload can not reliably be used for hdf5 files containing multiple datasets. Autoload works, but I am not sure how it determines which dataset to load. However, loading individual datasets from a hdf5 file works if the dataset name is provided.
Tests are included.
See also [issue 41](https://github.com/conradsnicta/armadillo-code/issues/41).

Remaining question: How should I provide documentation?

conradsnicta commented on include/armadillo_bits/arma_forward.hpp:
```diff
@@ -280,6 +280,7 @@ struct hdf5_name
   {
   const std::string filename;
   const std::string dsname;
+  const bool append = false;
```
This doesn't work with the older C++98 standard.  Armadillo currently supports C++98, C++11 and C++14.
The code needs to be changed to:
     
     const bool append;

and the `append` variable needs to be initialised explicitly in all constructors.

conradsnicta commented on include/armadillo_bits/diskio_meat.hpp:
```diff
@@ -1269,8 +1269,19 @@ diskio::save_hdf5_binary(const Mat<eT>& x, const hdf5_name& spec)
     // Set up the file according to HDF5's preferences
-    hid_t file = arma_H5Fcreate(tmp_name.c_str(), H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
+    hid_t file;
+    if ((file_is_hdf5 == 1) & (spec.append == true))
```
The single `&` means bitwise and operation, which is not correct in this instance.  It should be `&&` to indicate logical and operation.

conradsnicta commented on include/armadillo_bits/diskio_meat.hpp:
```diff
@@ -1319,7 +1331,10 @@ diskio::save_hdf5_binary(const Mat<eT>& x, const hdf5_name& spec)
+      }
     
     return save_okay;
     }
```
The `diskio::save_hdf5_binary(const Mat<eT>& x, const hdf5_name& spec)` function was extended to support appending.  The documentation for `.save()` applies to both matrices and cubes, so the corresponding function to handle cubes also needs to be extended: `diskio::save_hdf5_binary(const Cube<eT>& x, const hdf5_name& spec)`

conradsnicta commented:
> Remaining question: How should I provide documentation?

You can extend the `docs.html` file, and include the modifications as part of the pull request.
https://github.com/conradsnicta/armadillo-code/blob/unstable/docs.html

rcurtin commented on include/armadillo_bits/arma_forward.hpp:
```diff
@@ -291,6 +292,12 @@ struct hdf5_name
+  hdf5_name(const std::string& in_filename, const std::string& in_dsname, const bool& in_append)
+    : filename(in_filename)
+    , dsname  (in_dsname  )
+    , append  (in_append  )
```
I wonder, could you just use a default argument of `false` for `in_append` here and reduce the number of constructors?  I think even the `in_dsname` argument could have a default too.

rcurtin commented on include/armadillo_bits/diskio_meat.hpp:
```diff
@@ -1305,7 +1316,8 @@ diskio::save_hdf5_binary(const Mat<eT>& x, const hdf5_name& spec)
 
     hid_t dataset = arma_H5Dcreate(groups.size() == 0 ? file : groups[groups.size() - 1], dataset_name.c_str(), datatype, dataspace, H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
-    
+    if (dataset < 0) { arma_debug_warn("Mat::save(): dataset ", dataset_name.c_str(), std::string(" already exists in ") + spec.filename.c_str());} //can never happen if file is overwritten, so no need to use tmp_name.c_str()
```
Should this be `arma_check()` instead of `arma_debug_warn()`?  I think this should cause the program to fail in both debugging and non-debugging mode.

JWichmann commented:
Hi everyone,
thanks for the feedback. Sorry for the late fixes, I was traveling for a couple of days.
I fixed the issues mentioned above. Not sure about the C++98 issues as I could never get the tests to run with C++98 or C++03.
I did not change to arma_check but rather to the unconditional arma_warn.
Cubes can now be appended to hdf5 files as well.
I also added a test for saving a cube in hdf5 format.
Documentation will be written once my work here is finished and merged.

conradsnicta commented:
I'll merge this as is. There is also an issue with handling a hierarchy of groups, which I'll fix after the merge.
